### PR TITLE
Unwrap single status.MultiErrors

### PR DESCRIPTION
**Pull request by karlkfi**
- Fix Applier/Destroyer.addError from wrapping/appending nil errors
- Add status.Wrap for wrapping status.Errors without an existing MultiError.
- Fix status.Append to avoid wrapping errors that are already a MultiError
- Add testerrors.AssertEqual to simplify testing status.Errors. Ideally we would just fix statis.Error.Is, but that would require fixing all our test assertions. This approach is a bit easier, so we can fix assertions a little at a time to make them more strict.
- Backfilled unit tests for ErrorForResourceWithResource

### CODE REVIEW

**Review by nan-yu — APPROVED** [review]
/lgtm

File: pkg/applier/applier_test.go
Comment (sdowell) [inline pkg/applier/applier_test.go:142]: Does this still verify the `resources` fields? I don't think this is all surfaced by the `error` interface, so it seems like this might be tested - although I haven't yet dug into the `testutils` code from cli-utils

/hold

Comment (sdowell) [reply pkg/applier/applier_test.go:142]: This also reduces the test to basically asserting `foo() == foo()`, where `foo` is the real code path. The test assertions should be written such that `foo() == expectedValue`. Reading this test it's not really clear what the output should look like

Comment (karlkfi) [reply pkg/applier/applier_test.go:142]: The test here is asserting that Applier.Apply returns the error received from a specific event and wraps it appropriately with the resource.

Comment (sdowell) [reply pkg/applier/applier_test.go:142]: With this test I could change the underlying implementation of `ErrorForResourceWithResource` to no longer include the ResourceRef and the test would still pass

Comment (karlkfi) [reply pkg/applier/applier_test.go:142]: Yes, but this test isn't testing that ErrorForResourceWithResource exposes the right fields/methods. It's just testing that Applier.Apply returned the expected error. 

If you want to confirm ErrorForResourceWithResource has a ResourceRef field/method, that should be a test of ErrorForResourceWithResource itself.

Comment (sdowell) [reply pkg/applier/applier_test.go:142]: Can you add some individual tests for that then, since the coverage is being removed here?

Comment (karlkfi) [reply pkg/applier/applier_test.go:142]: Fine, but we have zero tests for error builders and dozens of usages of BuildWithResources already. ErrorForResourceWithResource just wraps BuildWithResources the same way everything else does.

Comment (sdowell) [reply pkg/applier/applier_test.go:142]: Agreed, which is why I asked for these assertions to be added to these tests in the first place. I find the error builders themselves pretty obfuscated and untested.

This bugfix was specifically for the ResourceRef fields on the RSync status, so I wanted to make sure there were assertions in place which verify the behavior and prevent regression

Comment (karlkfi) [reply pkg/applier/applier_test.go:142]: Done

Comment (sdowell) [reply pkg/applier/applier_test.go:142]: Thank you!

**Review by sdowell — APPROVED** [review]
/lgtm
/unhold

**Comment by karlkfi** [comment]
Fixed the e2e error. Need new LGTM

**Comment by karlkfi** [comment]
Fixed the e2e harder, for real this time.

**Pull request merged**
